src/Transfer.js

```diff
@@ -0,0 +1,390 @@
+import { LeftHeader } from './Transfer/LeftHeader.js'
+import { LeftSide } from './Transfer/LeftSide.js'
+import { PickedOptions } from './Transfer/PickedOptions.js'
+import { RemoveAll } from './Transfer/RemoveAll.js'
+import { RemoveIndividual } from './Transfer/RemoveIndividual.js'
+import { ReorderingActions } from './Transfer/ReorderingActions.js'
+import { RightFooter } from './Transfer/RightFooter.js'
+import { RightSide } from './Transfer/RightSide.js'
+import { SourceOptions } from './Transfer/SourceOptions.js'
+import {
+    addAllSelectableSourceOptions,
+    addIndividualSourceOptions,
+    addKeyToReactOptions,
+    createDoubleClickHandlers,
+    extractPickedReactOptions,
+    filterOptions as defaultFilterOptions,
+    filterOutOptions,
+    getSubsetByFilter,
+    isReorderDownDisabled,
+    isReorderUpDisabled,
+    moveHighlightedPickedOptionDown,
+    moveHighlightedPickedOptionUp,
+    removeAllPickedOptions,
+    removeIndividualPickedOptions,
+    useHighlightedOptions,
+} from './Transfer/helper/index.js'
+import {
+    singleSelectedPropType,
+    multiSelectedPropType,
+} from './common-prop-types.js'
+
+/**
+ * @module
+ * @param {Transfer.PropTypes} props
+ *
+ * @returns {React.Component}
+ *
+ * @example import { Transfer } from @dhis2/ui-core
+ * @see Specification: {@link https://github.com/dhis2/design-system/blob/master/organisms/transfer.md|Design system}
+ * @see Live demo: {@link /demo/?path=/story/transfer--basic|Storybook}
+ */
+export const Transfer = ({
+    onChange,
+
+    children,
+    className,
+    dataTest,
+    disabled,
+    emptyOptionsPlaceholder,
+    emptySelectionPlaceholder,
+    enableOrderChange,
+    filter,
+    filterLabel,
+    filterOptions,
+    height,
+    initialFilter,
+    labelAddAll,
+    labelAddIndividual,
+    labelRemoveAll,
+    labelRemoveIndivitual,
+    leftFooter,
+    leftHeader,
+    maxSelections,
+    optionsWidth,
+    rightFooter,
+    selected,
+    selectedWidth,
+    showFilterInput,
+    onFilterChange,
+}) => {
+    /**
+     * As this component will filter and sort children
+     * in the selected area, adding keys will prevent
+     * reconciliation from suffering badly
+     */
+    const allOptions = addKeyToReactOptions(children)
+
+    /**
+     * Used in the "Filter" section and for
+     * limiting the selectable source options
+     *
+     * Filter can be controlled & uncontrolled.
+     * Providing the "onFilterChange" callback will make it a controlled value
+     */
+    const isFilterEnabled = showFilterInput || onFilterChange
+    const [internalFilter, setInternalFilter] = useState(initialFilter)
+    const actualFilter = onFilterChange ? filter : internalFilter
+
+    /**
+     * These are all the available option react elements.
+     * It will replace all selected options with null
+     */
+    const sourceOptions = filterOutOptions(allOptions, selected)
+    const filteredSourceOptions = getSubsetByFilter({
```

Comment:
To me the wording here is slightly unclear. After reading what's going on, to me this would make sense:

```
const availableOptions = filterSelectedOptions(allOptions, selected)
const filteredAvailableOptions = getSubsetByFilter({ })
```

Technically, I'd also expect filterSelectedOptions to filter out the null values, but I'm not too bothered by that.